test/cpp/end2end/xds_end2end_test.cc

```diff
@@ -838,6 +839,7 @@ class AdsServiceImpl : public std::enable_shared_from_this<AdsServiceImpl> {
                         SubscriptionMap* subscription_map,
                         SentState* sent_state,
                         absl::optional<DiscoveryResponse>* response) {
+      grpc_core::MutexLock lock(&parent_->ads_mu_);
```

Comment:
This doesn't look right to me.  `ProcessRequest()` is called while already holding the lock, so I think this will deadlock.

I think you just need to tag this method with `ABSL_EXCLUSIVE_LOCKS_REQUIRED`.

Reply:
You're absolutely right. Done.